src/zcap.rs

```diff
@@ -0,0 +1,473 @@
+        }
+    }
+
+    // https://w3c-ccg.github.io/ld-proofs/
+    pub async fn generate_proof(
+        &self,
+        jwk: &JWK,
+        options: &LinkedDataProofOptions,
+    ) -> Result<Proof, Error> {
+        LinkedDataProofs::sign(self, options, jwk).await
+    }
+
+    /// Prepare to generate a linked data proof. Returns the signing input for the caller to sign
+    /// and then pass to [`ProofPreparation::complete`] to complete the proof.
+    pub async fn prepare_proof(
+        &self,
+        public_key: &JWK,
+        options: &LinkedDataProofOptions,
+    ) -> Result<ProofPreparation, Error> {
+        LinkedDataProofs::prepare(self, options, public_key).await
+    }
+
+    pub fn set_proof(self, proof: Proof) -> Self {
+        Self {
+            proof: Some(proof),
+            ..self
+        }
+    }
+}
+
+#[async_trait]
+impl<A, C> LinkedDataDocument for Delegation<A, C>
+where
+    A: Serialize + Send + Sync + Clone,
+    C: Serialize + Send + Sync + Clone,
+{
+    fn get_contexts(&self) -> Result<Option<String>, Error> {
+        Ok(Some(serde_json::to_string(&self.context)?))
+    }
+
+    async fn to_dataset_for_signing(
+        &self,
+        parent: Option<&(dyn LinkedDataDocument + Sync)>,
+    ) -> Result<DataSet, Error> {
+        let mut copy = self.clone();
+        copy.proof = None;
+        let json = serde_json::to_string(&copy)?;
+        let more_contexts = match parent {
+            Some(parent) => parent.get_contexts()?,
+            None => None,
+        };
+        let mut loader = StaticLoader;
+        json_to_dataset(&json, more_contexts.as_ref(), false, None, &mut loader).await
+    }
+}
+
+// limited initial definition of a ZCAP Invocation, generic over Action
+#[derive(Debug, Serialize, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Invocation<A> {
+    #[serde(rename = "@context")]
+    pub context: Contexts,
+    pub id: URI,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub capability_action: Option<A>,
+    // This field is populated only when using
+    // embedded proofs such as LD-PROOF
+    //   https://w3c-ccg.github.io/ld-proofs/
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub proof: Option<Proof>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(flatten)]
+    pub property_set: Option<Map<String, Value>>,
+}
+
+impl<A> Invocation<A>
+where
+    A: Serialize + Send + Sync + Clone,
+{
+    pub async fn verify<C>(
+        &self,
+        _options: Option<LinkedDataProofOptions>,
+        resolver: &dyn DIDResolver,
+        // TODO make this a list for delegation chains
+        target_capability: &Delegation<A, C>,
+    ) -> VerificationResult
+    where
+        C: Serialize + Send + Sync + Clone,
+    {
+        match &self.proof {
+            None => VerificationResult::error("No applicable proof"),
+            Some(proof) => {
+                let mut result = proof.verify(self, resolver).await;
```

Comment:
Should the proof's `proofPurpose` be checked and required to be `capabilityInvocation`?

Reply:
yes :+1: